### ISSUE TITLE
Trying to understand what's happening under the hood

### ISSUE THREAD

**Issue by janslu** [qbus-agentur/nginx_cache#2]
Hi, 
I'm trying to debug a problem on a multidomain tree. I have two issues:

1. clearing all cache doesn't seem to clear cache on other domain (it is configured properly, I log all purge requests, and I see requests for single urls, but not '*' calls). 

2. It seems that from time to time pages on the other domain become 'unpurgable' - I cannot force cache clearing on them - looks like the identifier is being removed from the cache database, but the purge call is not being made. I am logging all requests and I cannot grasp what is really happening. 

As a side question - any idea why I get repeated tags in cache set queries? When I log on line 64 in SetPageCacheHook.php just under `($cachable) {`
I get the following tags array: 
```
(
    [0] => pageId_6069
    [1] => tx_news
    [2] => tx_news_pid_7475
    [3] => tx_news_pid_7475
    [4] => pageId_6069
)
```
Does it change anything? Is this something I should look into?

Thanks for your work, I use the nginx_cache to speed our site considerably, it's just that sometimes it seems to work "too well"...

**Comment by bnf** [comment]
> 1. clearing all cache doesn't seem to clear cache on other domain (it is configured properly, I log all purge requests, and I see requests for single urls, but not '*' calls).

Upfront: flushing all caches via CLI is currently not supported (will be in a future version using an `ext_conf_template.txt` option where a base URL needs to be defined) – but I suppose you flush from the TYPO3 Backend.

Upfront #2: You've configured nginx to use our `purge.pm` perl script? (Not something like nginx plus, or so?)

If you see no '*' calls, then you may debug if `$url` has a sensible value in https://github.com/qbus-agentur/nginx_cache/blob/b0e7c0ec1220a6253c2d7cf5e4d2dc06195509e3/Classes/Cache/Backend/NginxCacheBackend.php#L90
You may also try if all caches are flushed when you invoke `curl -X PURGE '$url'` (replacing `$url` with the value). `$url` should be `https://your-domain.tld/*`.

> 2. It seems that from time to time pages on the other domain become 'unpurgable' - I cannot force cache clearing on them - looks like the identifier is being removed from the cache database, but the purge call is not being made. I am logging all requests and I cannot grasp what is really happening.

Haven't had that problem here. So we need some debugging here..

How do you force cache clearing on individual pages? I suppose you use the flush button for the specific page in the backend? That means TYPO3 flushes the page cache by clearing a tag like `pageId_{uid}`.
Two things that may go wrong:

- The tag `pageId_{uid}` may be missing (unlikely, but who knows..)
- Maybe the cache entry is considered invalid by the TYPO3 database cache backend, but not by nginx (e.g. through timezone issues). (In that case the cache entry would not be found and thus no PURGE request to the nginx server performed)

You can list your cache entries using the following SQL and check whether the unpurgable URLs missing proper cache tags (`pageId_{uid}`) or have an `expires` value in the past:
```
select content, FROM_UNIXTIME(expires) as expires, cf_nginx_cache.identifier, GROUP_CONCAT(tag) as tags from cf_nginx_cache, cf_nginx_cache_tags where  cf_nginx_cache.identifier = cf_nginx_cache_tags.identifier GROUP BY cf_nginx_cache.identifier;
```


> As a side question - any idea why I get repeated tags in cache set queries?

Multiple cache tags are not harmful (besides wasting storage). I experience the same on my installations. I have never tracked that down, but I think that happens due to `TypoScriptFrontendController::setPageCacheContent()` being called twice per page generation: First with temporary content ('Page is being generated') and then with with the real content. With each invocation that cache tag is added with `$this->pageCacheTags[] = 'pageId_' . $cacheData['page_id'];` – so a core bug.

You can see that the core caches are affected as well:
```
select GROUP_CONCAT(tag), identifier from cf_cache_pages_tags group by identifier;
```

**Comment by janslu** [comment]
Thank you for your time. I've been debugging this for the past few hours, and I'm just amazed at how little I knew about anything. 

My main problems seems to be caused by the fact that I'm using RedisCache as a page cache. I guess this causes the purge hook for single pages not to launch at all when using the button in the BE to clear cache of a page. This is for all domains. Unfortunately the identificator is being removed from the nginx_cache and nginx_cache_tags tables, but purge calls never happen.

Mass clearing cache for all pages using BE lightning menu will successfully clear cache but only for TYPO3_SITE_URL, which is my first domain. Not for the second one. Only after I added manually the second domain in the flush function, the cache is being cleared properly. 

SOME pages have their caches cleared by purge calls but this must be something like template changes, or cache timeout coming from typo3. That's why I do have some purge calls logged in my nginx logs. 

Thanks for the tags information - when debugging this seemed suspicious. 

I will dive deeper into the redis cache classes to check if I can safely override them. But it seems that it is necessary to iterate over domains when clearing all cache in the BE. I will look into it later on, but so far I have no idea where typo3 stores this information...

**Comment by bnf** [comment]
I'm pretty sure this has been fixed by #15.
The scenario you describe happens when the database cache is cleared without a PURGE request, which unfortunately happened when `flushByTags()` was triggered to individual pages in TYPO3 >= v8.

Fix released in v2.2.3.
